### PR TITLE
random-util: add a longer comment explaining our RDRAND use

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/basic/random-util.c [hunk 1]
```diff
@@ -33,6 +33,66 @@
...
+         *           hash functions for its hash tables, with a seed generated randomly. The hash tables
+         *           systemd employs watch the fill level closely and reseed if necessary. This allows use of
+         *           a low quality RNG initially, as long as it improves should a hash table be under attack:
+         *           the attacker after all needs to to trigger many collisions to exploit it for the purpose
+         *           of DoS, but if doing so improves the seed the attack surface is reduced as the attack
+         *           takes place.
+         *
+         * Some cases where we do NOT use RDRAND are:
+         *
+         *         • Generation of cryptographic key material 🔑
```
Comment (CameronNemo) [inline src/basic/random-util.c:83]: Is this unicode? Why?

Comment (poettering) [reply src/basic/random-util.c:83]: We generally use UTF-8 encoding in our source files.

File: src/basic/random-util.c [hunk 2]
```diff
@@ -33,6 +33,66 @@
 
 int rdrand(unsigned long *ret) {
 
+        /* So, you are a "security researcher", and you wonder why we bother with using raw RDRAND here,
+         * instead of sticking to /dev/urandom or getrandom()?
```
Comment (mbiebl) [inline src/basic/random-util.c:37]: `you are a "security researcher", and ` 
I would just drop that. It's unnecessary confrontional and doesn't really help getting the message across.

**Comment by mbiebl** [comment]
Is there a good reason to use RDRAND beyond early boot? Shouldn't we just switch to `getrandom()` once we have enough entropy, assuming we can determine when it is "safe" to use `getrandom()`?

**Comment by poettering** [comment]
> Is there a good reason to use RDRAND beyond early boot? Shouldn't we just switch to `getrandom()` once we have enough entropy, assuming we can determine when it is "safe" to use `getrandom()`?

I am not sure how we would communicate that sanely (because we need it not only in PID 1 but in all our processes), or how to even define "early boot"... And I must say not trusting your CPU to work at all is a very uncomfortable place to be in. We have to trust it ultimately, since it directly executes our code.

If the RDRAND is concerning you, please work with the CPU vendor to ensure this gets fixed/worked around in firmware, the kernel, elsewhere. But I don't think stopping to use RDRAND is warranted. 

In general I think it's a good idea to push RDRAND into the common codepaths, so that we detect issues like that. If RDRAND is just the codepath that is used in a corner case we would never have noticed the issue, and that's worse than things are now, as the failures would have been extremely spurious. by preferring RDRAND over getrandom() where we are fine with it we know that both options are used frequently, so we know everything is cool with them and can be sure both codepaths are regularly validated.

File: src/basic/random-util.c [hunk 3]
```diff
@@ -33,6 +33,66 @@
...
+         * getrandom() is generally not initialized yet. It is very common that initialization of the random
+         * pool takes a longer time (up to many minutes), in particular on embedded devices that have no
+         * explicit hardware random generator, as well as in virtualized environments such as major cloud
+         * installations that do not provide virtio-rng or a similar mechanism.
+         *
+         * In such an environment using getrandom() synchronously means we'd block the entire system boot-up
+         * until the pool is initialized, i.e. *very* long. Using getrandom() asynchronously (GRND_NONBLOCK)
+         * would mean acquiring randomness during early boot would simply fail. Using /dev/urandom would mean
+         * generating many kmsg log messages about our use of it before the random pool is properly
+         * initialized. Neither of these outcomes is desirable.
```
Comment (keszybz) [inline src/basic/random-util.c:49]: Neither only works with two alternatives, but OK.

File: src/basic/random-util.c [hunk 4]
```diff
@@ -276,6 +351,38 @@ void pseudo_random_bytes(void *p, size_t n) {
 
 void random_bytes(void *p, size_t n) {
 
+        /* This returns high quality randomness if we can get it cheaply. If we can't because for some reason
+         * it is not available we'll try some crappy fallbacks.
```
Comment (keszybz) [inline src/basic/random-util.c:355]: "because for some reason it is not available" → this contains no informational value.

**Pull request merged**